File: app/src/main/java/james/metronome/views/SeekBar.java
package james.metronome.views;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;

import com.afollestad.aesthetic.Aesthetic;

import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Consumer;
import me.jfenn.androidutils.DimenUtils;

public class SeekBar extends View implements View.OnTouchListener {

    private Paint paint;
    private Paint secondaryPaint;
    private Paint accentPaint;

    private OnProgressChangeListener listener;
    private int progress;
    private int maxProgress = 100;
    private float touchDiff;

    private float strokeWidth = DimenUtils.dpToPx(2);

    private Disposable textColorPrimarySubscription;
    private Disposable textColorSecondarySubscription;
    private Disposable colorAccentSubscription;

    public SeekBar(Context context) {
        this(context, null);
    }

    public SeekBar(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public SeekBar(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        paint = new Paint();
        paint.setColor(Color.BLACK);
        paint.setStrokeWidth(strokeWidth);
        paint.setAntiAlias(true);

        secondaryPaint = new Paint();
        secondaryPaint.setColor(Color.BLACK);
        secondaryPaint.setStrokeWidth(strokeWidth);
        secondaryPaint.setAntiAlias(true);

        accentPaint = new Paint();
        accentPaint.setColor(Color.BLACK);
        accentPaint.setStrokeWidth(strokeWidth);
        accentPaint.setAntiAlias(true);

        setOnTouchListener(this);
        setClickable(true);

        subscribe();
    }

    public void subscribe() {
        textColorPrimarySubscription = Aesthetic.Companion.get()
                .textColorPrimary()
                .subscribe(new Consumer<Integer>() {
                    @Override
                    public void accept(Integer integer) throws Exception {
                        paint.setColor(integer);
                        invalidate();
                    }
                });

        textColorSecondarySubscription = Aesthetic.Companion.get()
                .textColorSecondary()
                .subscribe(new Consumer<Integer>() {
                    @Override
                    public void accept(Integer integer) throws Exception {
                        secondaryPaint.setColor(integer);
                        invalidate();
                    }
                });

        colorAccentSubscription = Aesthetic.Companion.get()
                .colorAccent()
                .subscribe(new Consumer<Integer>() {
                    @Override
                    public void accept(Integer integer) throws Exception {
                        accentPaint.setColor(integer);
                        invalidate();
                    }
                });
    }

    public void unsubscribe() {
        textColorPrimarySubscription.dispose();
        textColorSecondarySubscription.dispose();
        colorAccentSubscription.dispose();
    }

    public void setProgress(int progress) {
        this.progress = progress;
        invalidate();

        if (listener != null)
            listener.onProgressChange(progress);
    }

    public int getProgress() {
        return progress;
    }

    public void setMaxProgress(int maxProgress) {
        this.maxProgress = maxProgress;
        invalidate();
    }

    public void setOnProgressChangeListener(OnProgressChangeListener listener) {
        this.listener = listener;
    }

    @Override
    public boolean onTouch(View view, MotionEvent event) {
        float x = event.getX();
        switch (event.getAction()) {
            case MotionEvent.ACTION_DOWN:
                touchDiff = x - (((float) progress / maxProgress) * view.getMeasuredWidth());
            case MotionEvent.ACTION_MOVE:
                int progress = (int) (maxProgress * ((x - touchDiff) / view.getMeasuredWidth()));
                if (progress < 0)
                    progress = 0;
                else if (progress > maxProgress)
                    progress = maxProgress;

                setProgress(progress);
        }
        return false;
    }

    public interface OnProgressChangeListener {
        void onProgressChange(int progress);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        secondaryPaint.setAlpha(255);
        canvas.drawLine(0, strokeWidth/2, canvas.getWidth(), strokeWidth/2, secondaryPaint);

        int currentWidth = (int) (canvas.getWidth() * ((float) progress / maxProgress));
        for (int i = 0; i < maxProgress; i += 10) {
            int width = (int) (canvas.getWidth() * ((float) i / maxProgress));
            secondaryPaint.setAlpha(Math.max(255 - (int) ((float) Math.abs(width - currentWidth) * 1000 / canvas.getWidth()), 0));
            canvas.drawLine(width, strokeWidth/2, width, DimenUtils.dpToPx(i % 20 == 0 ? 14 : 8), secondaryPaint);
        }

        canvas.drawLine(0, strokeWidth/2, currentWidth, strokeWidth/2, accentPaint);
        canvas.drawLine(currentWidth, strokeWidth/2, currentWidth, DimenUtils.dpToPx(18), accentPaint);
    }
}

File: app/src/main/java/me/jfenn/metronome/views/SeekBar.kt
package me.jfenn.metronome.views

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View
import android.view.View.OnTouchListener
import me.jfenn.androidutils.dpToPx
import me.jfenn.metronome.R
import me.jfenn.metronome.utils.getThemedColor
import kotlin.math.abs

class SeekBar @JvmOverloads constructor(
        context: Context?,
        attrs: AttributeSet? = null,
        defStyleAttr: Int = 0
) : View(context, attrs, defStyleAttr), OnTouchListener {

    private val lineWidth = dpToPx(2f).toFloat()

    private val secondaryPaint = Paint().apply {
        color = context?.getThemedColor(R.attr.textColorPrimary) ?: Color.BLACK
        strokeWidth = lineWidth
        isAntiAlias = true
    }

    private val accentPaint = Paint().apply {
        color = context?.getThemedColor(R.attr.colorAccent) ?: Color.RED
        strokeWidth = lineWidth
        isAntiAlias = true
    }

    private var listener: OnProgressChangeListener? = null
    private var progress = 0
    private var displayedProgress = 0f
    private var maxProgress = 100
    private var touchDiff = 0f

    init {
        setOnTouchListener(this)
        isClickable = true
    }

    fun setProgress(progress: Int, immediate: Boolean = false) {
        this.progress = progress
        if (immediate) this.displayedProgress = progress.toFloat()

        listener?.onProgressChange(progress)
        invalidate()
    }

    fun getProgress(): Int {
        return progress
    }

    fun setMaxProgress(maxProgress: Int) {
        this.maxProgress = maxProgress
        invalidate()
    }

    fun setOnProgressChangeListener(listener: OnProgressChangeListener?) {
        this.listener = listener
    }

    override fun onTouch(view: View, event: MotionEvent): Boolean {
        val x = event.x
        when (event.action) {
            MotionEvent.ACTION_DOWN -> {
                touchDiff = x - progress.toFloat() / maxProgress * view.measuredWidth
                var progress = (maxProgress * ((x - touchDiff) / view.measuredWidth)).toInt()
                if (progress < 0) progress = 0 else if (progress > maxProgress) progress = maxProgress
                setProgress(progress, true)
            }
            MotionEvent.ACTION_MOVE -> {
                var progress = (maxProgress * ((x - touchDiff) / view.measuredWidth)).toInt()
                if (progress < 0) progress = 0 else if (progress > maxProgress) progress = maxProgress
                setProgress(progress, true)
            }
        }
        return false
    }

    interface OnProgressChangeListener {
        fun onProgressChange(progress: Int)
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        secondaryPaint.alpha = 255
        canvas.drawLine(0f, lineWidth / 2, width.toFloat(), lineWidth / 2, secondaryPaint)
        val currentWidth = (width * (displayedProgress / maxProgress)).toInt()
        var i = 0
        while (i < maxProgress) {
            val width = (width * (i.toFloat() / maxProgress)).toInt()
            secondaryPaint.alpha = (255 - (abs(width - currentWidth).toFloat() * 1000 / this.width).toInt()).coerceAtLeast(0)
            canvas.drawLine(width.toFloat(), lineWidth / 2, width.toFloat(), dpToPx(if (i % 20 == 0) 14f else 8f).toFloat(), secondaryPaint)
            i += 10
        }
        canvas.drawLine(0f, lineWidth / 2, currentWidth.toFloat(), lineWidth / 2, accentPaint)
        canvas.drawLine(currentWidth.toFloat(), lineWidth / 2, currentWidth.toFloat(), dpToPx(18f).toFloat(), accentPaint)

        if (displayedProgress != progress.toFloat()) {
            displayedProgress = if (abs(displayedProgress - progress) < 0.5f)
                progress.toFloat()
            else (displayedProgress + progress) / 2f

            postInvalidate()
        }
    }
}